subsys/bluetooth/services/rscs.c

```diff
@@ -0,0 +1,431 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/printk.h>
+#include <sys/byteorder.h>
+#include <zephyr.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/hci.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/uuid.h>
+#include <bluetooth/gatt.h>
+
+#include <bluetooth/services/rscs.h>
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(bt_rscs, CONFIG_BT_RSCS_LOG_LEVEL);
+
+/* RSC Application error codes */
+#define RSC_ERR_IN_PROGRESS		0x80
+#define RSC_ERR_CCC_CONFIG		0x81
+
+/* SC Control Point Opcodes */
+#define SC_CP_OP_SET_CUMULATIVE		0x01
+#define SC_CP_OP_CALIBRATION		0x02
+#define SC_CP_OP_UPDATE_LOC		0x03
+#define SC_CP_OP_REQ_SUPP_LOC		0x04
+#define SC_CP_OP_RESPONSE		0x10
+
+/* SC Control Point Response Values */
+#define SC_CP_RSP_SUCCESS		0x01
+#define SC_CP_RSP_OP_NOT_SUPP		0x02
+#define SC_CP_RSP_INVAL_PARAM		0x03
+#define SC_CP_RSP_FAILED		0x04
+
+/* RSC Measurement Flags */
+#define RSC_INSTANT_STRIDE_LEN_PRESENT	BIT(0)
+#define RSC_TOTAL_DISTANCE_PRESENT	BIT(1)
+#define RSC_WALKING_OR_RUNNING_BIT	BIT(2)
+
+/* Running Speed and Cadence Service Feature bits */
+#define RSC_FEAT_INSTA_STRIDE_LEN	BIT(0)
+#define RSC_FEAT_TOTAL_DISTANCE		BIT(1)
+#define RSC_FEAT_WALKING_RUNNING	BIT(2)
+#define RSC_FEAT_SENSOR_CALIB_PROC	BIT(3)
+#define RSC_FEAT_MULTI_SENSORS		BIT(4)
+
+/* Attribute indexes */
+#define RSC_SVC_MEAS_ATTR_IDX		1
+#define RSC_SVC_CTRL_POINT_ATTR_IDX	8
+
+static struct {
+	uint8_t num_location;
+	bool meas_notify_enabled;
+	bool ctrl_pt_indicate_enabled;
+	bool indicating;
+	bt_rscs_evt_handler evt_handler;
+	const struct bt_rscs_cp_cb *cp_cb;
+	struct bt_rscs_features features;
+	enum bt_rscs_location sensor_loc;
+	enum bt_rscs_location supported_locations[RSC_LOC_AMT];
+	struct bt_gatt_indicate_params ind_params;
+} rscs_inst;
```

Comment:
I would prefer to sort this structure from biggest type to lowest but this is only a suggestion.

Reply:
I reordered the structure fields to optimize memory.